hardware/VirtualThermostat.cpp

```diff
@@ -0,0 +1,682 @@
+	float RoomSetPointTemp = GetSetPointTemperature(devIdx);
+
+	return GetMode(RoomSetPointTemp, EcoTemp, ConforTemp);
+};
+
+std::string VirtualThermostat::GetMode(float curTemp, float EcoTemp, float ConfTemp)
+{
+	float  moy = (EcoTemp + ConfTemp) / 2;
+	if (curTemp <= moy)  return ThermostatModeIntToString(Eco);
+	else               return ThermostatModeIntToString(Confor);
+}
+
+bool VirtualThermostat::SetThermostatState(const std::string& idx, const int newState)
+{
+	VirtualThermostatMode mode = (VirtualThermostatMode)newState;
+	if (mode >= EndMode)
+		return false;
+	if (mode == Confor)
+		m_mainworker.SetSetPoint(idx, (float)m_VirtualThermostat->GetConfortTemp(idx.c_str()));
+	else   if (mode == Eco)
+		m_mainworker.SetSetPoint(idx, (float)m_VirtualThermostat->GetEcoTemp(idx.c_str()));
+	else   if (mode == FrostProtection)
+		m_mainworker.SetSetPoint(idx, (float)ConvertTemperatureUnit((float)TEMPERATURE_HG));
+	else   if (mode == Off)
+		m_mainworker.SetSetPoint(idx, (float)ConvertTemperatureUnit((float)TEMPERATURE_OFF));
+	return true;
+}
+
+//return the thermostat room temperature 
+std::string VirtualThermostat::GetRoomTemperature(std::string& devIdx)
+{
+	std::string temp = GetOptionFromDeviceId(devIdx, "RoomTemp");
+	return (temp);
+};
+
+std::string VirtualThermostat::GetSetPoint(std::string& devIdx)
+{
+	return (m_sql.GetDeviceValue("svalue", devIdx.c_str()).c_str());
+}
+
+float VirtualThermostat::GetSetPointTemperature(std::string& devIdx)
+{
+	return ((float)std::stof(GetSetPoint(devIdx).c_str()));
+};
+
+//option is not in base64
+std::string VirtualThermostatGetOption(const std::string optionName, const std::string& options, const bool decode)
+{
+	//if option present
+	TOptionMap Option = m_sql.BuildDeviceOptions(options, decode);
+	return (Option[optionName]);
+}
+//return temperature value from Svalue : is code temperature;humidity;???
+float VirtualThermostat::getTemperatureFromSValue(const char* sValue)
+{
+	std::vector<std::string> splitresults;
+	StringSplit(sValue, ";", splitresults);
+	if (splitresults.size() < 1)
+		return 0;
+	else
+		return (float)std::stof(splitresults[0].c_str());
+}
+//convert from  celcius to farenheit if unit=farenheit
+double  VirtualThermostat::ConvertTemperatureUnit(double tempcelcius)
+{
+	if (m_sql.m_tempunit == TEMPUNIT_F)
+	{
+		//Convert back to celcius
+		tempcelcius = ConvertToFahrenheit(tempcelcius);
+	}
+	return tempcelcius;
+}
+bool VirtualThermostat::GetLastValue(const char* DeviceID, int& nValue, std::string& sValue, struct tm& LastUpdateTime)
+{
+	bool result = false;
+	char szTmp[400];
+	std::string sLastUpdate;
+	//std::string sValue;
+	//struct tm LastUpdateTime;
+	time_t now = mytime(NULL);
+	struct tm tm1;
+	localtime_r(&now, &tm1);
+
+	sprintf(szTmp, "SELECT nValue,sValue,LastUpdate FROM DeviceStatus WHERE ( ID=%s ) ", DeviceID);
+	auto sqlresult = m_sql.safe_query(szTmp);
+
+	if (!sqlresult.empty())
+	{
+		nValue = (int)std::stoi(sqlresult[0][0].c_str());
```

Comment:
you do not need .c_str() here, also the cast to (int) is not needed

Reply:
Ok done.